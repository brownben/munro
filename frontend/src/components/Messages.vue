<!--
  Messages

  Messages are displayed at the bottom of the screen. On enter and exit they slide and fade in. It monitors the
    message store and displays the messages
-->

<template>
  <transition name="fade">
    <div
      is="transition-group"
      v-show="messages.length > 0"
      id="messages"
      name="list"
      mode="out-in"
      class="fixed bottom-0 right-0"
    >
      <p
        v-for="message of messages"
        :key="message.id"
        class="card-color m-3 font-heading z-40"
        @click="clear(message.id)"
      >
        {{ message.text }}
      </p>
    </div>
  </transition>
</template>

<script>
export default {
  name: 'Messages',

  data: function() {
    return {
      messages: this.$messages.messages,
    }
  },

  methods: {
    clear: function(id) {
      this.$messages.removeMessage(id)
    },
  },
}
</script>

<style scoped>
.list-enter-active,
.list-leave-active {
  transition: all 0.5s ease-out;
}
.list-enter,
.list-leave-to {
  opacity: 0;
  transform: translateX(10px);
}
.fade-enter-active,
.fade-leave-active {
  transition: opacity 0.5s;
}

.fade-enter,
.fade-leave-to {
  opacity: 0;
}
</style>
