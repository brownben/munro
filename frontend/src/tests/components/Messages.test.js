/*
  Messages Component Unit Tests
*/

import { mount } from '@vue/test-utils'
import Messages from '../../components/Messages'

test('Is a Vue Instance', () => {
  const wrapper = mount(Messages, {
    mocks: {
      $messages: { messages: [] },
    },
  })
  expect(wrapper.isVueInstance()).toBeTruthy()
})

test('Renders Correctly', () => {
  const messages = {
    messages: [
      { id: 0, text: 'Message 0' },
      { id: 1, text: 'Message 1' },
      { id: 2, text: 'Message 2' },
    ],
  }
  const wrapper = mount(Messages, {
    mocks: {
      $messages: messages,
    },
  })
  expect(wrapper.element).toMatchSnapshot()
})

test('Correct Number of Messages Displayed - 3 Messages', () => {
  const messages = {
    messages: [
      { id: 0, text: 'Message 0' },
      { id: 1, text: 'Message 1' },
      { id: 2, text: 'Message 2' },
    ],
  }
  const wrapper = mount(Messages, {
    mocks: {
      $messages: messages,
    },
  })
  expect(wrapper.findAll('p').length).toBe(3)
})

test('Correct Number of Messages Displayed - No Messages', () => {
  const messages = {
    messages: [],
  }
  const wrapper = mount(Messages, {
    mocks: {
      $messages: messages,
    },
  })
  expect(wrapper.findAll('p').length).toBe(0)
})

test('Clear Message when Clicked', () => {
  const removeMessage = jest.fn()
  const messages = {
    removeMessage: removeMessage,
    messages: [
      { id: 0, text: 'Message 0' },
    ],
  }
  const wrapper = mount(Messages, {
    mocks: {
      $messages: messages,
    },
  })
  wrapper.find('p').trigger('click')
  expect(removeMessage).toHaveBeenCalledTimes(1)
  expect(removeMessage).toHaveBeenLastCalledWith(expect.any(Number))
})
