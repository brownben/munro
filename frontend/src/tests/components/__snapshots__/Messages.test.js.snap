// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders Correctly 1`] = `
<div
  id="messages"
  mode="out-in"
  name="fade"
>
  <p>
    Message 0
  </p>
  <p>
    Message 1
  </p>
  <p>
    Message 2
  </p>
</div>
`;
